### PR TITLE
Add BCH Support

### PR DESCRIPTION
**Pull request by mainnet-pat**
Hi, this PR is intended to track the progress of bitcoincash integration into BasicSwap DEX, I'd appreciate you reviewing it and helping me navigate through the integration process.

As BCH is not a segwit coin, it can not be simply plugged into the existing protocol, like for example LTC. BCH-XMR swaps follow their own protocol, see https://gitlab.com/0353F40E/cross-chain-swap-ves. It employs a creating a swaplock smart contract which is a covenant to restrict spending to two paths - refund and swap. Swap path requires the OT-VES signature for XMR side to construct the spending key. More details in the link above. My work so far already contains the code required to construct the smart contract and spend it, so I am working further on building blocks required to implement the protocol in BSX.

Thanks in advance.

### CODE REVIEW

**Comment by mainnet-pat** [comment]
This week's activities were to continue evolving BCH interface to be compatible with XMR_SWAP protocol, (re-)implementing the segwit script and transaction building on a BCH non-segwit chain. In addition I have done some investigation and BCH seems to be a viable swap pair (follower) for segwit coins too if treated as an unscriptable coin.

**Comment by mainnet-pat** [comment]
This week's report features progress on further work towards implementing the BCH swap interface. Major achievement is the happy path (lock-spend) BCH-XMR swap completion, yay! More work is needed on three other paths:
* lock-refund
* lock-refund-spend
* lock-refund-refund

These will be easier to implement as most fundamentals are already in place.

**Comment by A60AB5450353F40E** [comment]
> I have done some investigation and BCH seems to be a viable swap pair (follower) for segwit coins too, if treated as an unscriptable coin.

Why does it need to be treated as unscriptable? "classic" [HTLC](https://github.com/tecnovert/basicswap/blob/master/basicswap/protocols/atomic_swap_1.py#L29) works just fine with BCH, too.

Because BCH has relatively much cheaper fees, maybe BCH could be the "lead" and BTC side treated as if unscriptable to reduce on-chain footprint, in which case you'd use the same BCH-XMR contract and adaptor sig (but no need dleq since it is same curve).

**Comment by mainnet-pat** [comment]
> Why does it need to be treated as unscriptable? "classic" [HTLC](https://github.com/tecnovert/basicswap/blob/master/basicswap/protocols/atomic_swap_1.py#L29) works just fine with BCH, too.
> 
> Because BCH has relatively much cheaper fees, maybe BCH could be the "lead" and BTC side treated as if unscriptable to reduce on-chain footprint, in which case you'd use the same BCH-XMR contract and adaptor sig (but no need dleq since it is same curve).

This is a great question, but it shifts this PR into another plane of discussion. My primary goal is to implement BCH-XMR swaps here

**Comment by mainnet-pat** [comment]
Hey guys, I am happy to drop the [WIP] tag from this PR. I have rigorously tested my changes and changed some gui code to display the BCH asset info correctly. Please check and verify.

This concludes my involvement as discussed with ofrnxmr. But I'd try to find more time if any issues are found.

File: basicswap/bin/run.py [hunk 1]
```diff
@@ -214,7 +214,7 @@ def runClient(fp, data_dir, chain, start_only_coins):
             if c in ('monero', 'wownero'):
                 if v['manage_daemon'] is True:
                     swap_client.log.info(f'Starting {display_name} daemon')
-                    filename = c + 'd' + ('.exe' if os.name == 'nt' else '')
+                    filename = (c if not c == "bitcoincash" else "bitcoin") + 'd' + ('.exe' if os.name == 'nt' else '')
```
Comment (nahuhh) [inline basicswap/bin/run.py:217]: Revert (wrong section)

File: basicswap/bin/prepare.py [hunk 2]
```diff
@@ -1755,6 +1784,19 @@ def main():
             'conf_target': 2,
             'core_version_group': 22,
         },
+        'bitcoincash': {
+            'connection_type': 'rpc' if 'bitcoincash' in with_coins else 'none',
+            'manage_daemon': True if ('bitcoincash' in with_coins and BCH_RPC_HOST == '127.0.0.1') else False,
```
Comment (nahuhh) [inline basicswap/bin/prepare.py:1789]:
```suggestion
            'connection_type': 'rpc',
            'manage_daemon': shouldManageDaemon('BCH'),
```

File: basicswap/chainparams.py [hunk 3]
```diff
@@ -432,7 +433,48 @@ class Coins(IntEnum):
...
+        'name': 'bitcoincash',
+        'ticker': 'BCH',
+        'message_magic': 'Bitcoin Signed Message:\n',
+        'blocks_target': 60 * 2,
+        'decimal_places': 8,
+        'has_cltv': True,
+        'has_csv': True,
+        'has_segwit': False,
+        'mainnet': {
+            'rpcport': 8332,
```
Comment (nahuhh) [inline basicswap/chainparams.py:447]: Conflict with btc

File: basicswap/chainparams.py [hunk 4]
```diff
@@ -432,7 +433,48 @@ class Coins(IntEnum):
...
+            'pubkey_address': 0,
+            'script_address': 5,
+            'key_prefix': 128,
+            'hrp': 'bitcoincash',
+            'bip44': 0,
+            'min_amount': 1000,
+            'max_amount': 100000 * COIN,
+        },
+        'testnet': {
+            'rpcport': 18332,
```
Comment (nahuhh) [inline basicswap/chainparams.py:457]: Same

File: basicswap/chainparams.py [hunk 5]
```diff
@@ -432,7 +433,48 @@ class Coins(IntEnum):
...
+            'script_address': 196,
+            'key_prefix': 239,
+            'hrp': 'bchtest',
+            'bip44': 1,
+            'min_amount': 1000,
+            'max_amount': 100000 * COIN,
+            'name': 'testnet3',
+        },
+        'regtest': {
+            'rpcport': 18443,
```
Comment (nahuhh) [inline basicswap/chainparams.py:468]: Same

File: basicswap/bin/prepare.py [hunk 6]
```diff
@@ -837,7 +861,6 @@ def prepareCore(coin, version_data, settings, data_dir, extra_opts={}):
                 verified = gpg.verify_file(fp, assert_path)
 
     ensureValidSignatureBy(verified, signing_key_name)
-
```
Comment (nahuhh) [inline basicswap/bin/prepare.py:840]: Whitespace

File: basicswap/static/js/offerstable.js [hunk 7]
```diff
@@ -166,7 +167,7 @@ const itemsPerPage = 100;
 const coinIdToName = {
   1: 'particl', 2: 'bitcoin', 3: 'litecoin', 4: 'decred',
   6: 'monero', 7: 'particl blind', 8: 'particl anon',
-  9: 'wownero', 11: 'pivx', 13: 'firo'
+  9: 'wownero', 11: 'pivx', 13: 'firo', 17: 'bitcoincash'
```
Comment (nahuhh) [inline basicswap/static/js/offerstable.js:170]:
```suggestion
  9: 'wownero', 11: 'pivx', 13: 'firo', 17: 'bitcoin cash'
```

**Pull request merged**